bothan-core/src/registry.rs

```diff
@@ -0,0 +1,108 @@
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::marker::PhantomData;
+
+use serde::{Deserialize, Serialize};
+
+use crate::registry::signal::Signal;
+
+pub mod post_processor;
+pub mod processor;
+pub mod signal;
+pub mod source;
+
+#[derive(Clone, Debug)]
+pub struct Invalid;
+
+#[derive(Clone, Debug)]
+pub struct Valid;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Registry<State = Invalid> {
+    #[serde(flatten)]
+    inner: HashMap<String, Signal>,
+    #[serde(skip_serializing, default)]
+    _state: PhantomData<State>,
+}
+
+impl Registry<Invalid> {
+    pub fn validate(self) -> Result<Registry<Valid>, Self> {
```

Comment:
can you document what needs to be validate and the algorithm? how do you make sure this function does not infinite loop if there are circular dependencies? need @RogerKSI @taobun @colmazia to help validate the algo

Reply:
+ this function should error with useful message if registry is not valid